### PR TITLE
Support gas estimate for pre eip1559

### PR DESCRIPTION
**Pull request by yrliou**
Resolves https://github.com/brave/brave-browser/issues/18167
Resolves https://github.com/brave/brave-browser/issues/17696

Screen Shot 2021-09-16 at 11 27 18 PM
Sending eth will use default value (21000 * 150 Gwei)

Screen Shot 2021-09-16 at 11 55 34 PM
This one is calculated by using values returned by eth_estimateGas & eth_gasPrice.
gas limit here is 0x604b, gas price here is 0xddbd3baad wei.

## Submitter Checklist:

- [ ] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [ ] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [ ] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [ ] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [ ] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [ ] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

**Comment by SIRAJUDHEENyakamoz** [comment]
@sirajudheen subair

File: components/brave_wallet/browser/eth_json_rpc_controller.cc [hunk 1]
```diff
@@ -485,4 +485,64 @@ GURL EthJsonRpcController::GetBlockTrackerUrlFromNetwork(std::string chain_id) {
...
+}
+
+void EthJsonRpcController::GetGasPrice(GetGasPriceCallback callback) {
+  auto internal_callback =
+      base::BindOnce(&EthJsonRpcController::OnGetGasPrice,
+                     weak_ptr_factory_.GetWeakPtr(), std::move(callback));
+  return Request(eth_gasPrice(), true, std::move(internal_callback));
+}
+
+void EthJsonRpcController::OnGetGasPrice(
```
Comment (bbondy) [inline components/brave_wallet/browser/eth_json_rpc_controller.cc:529]: Noting for a future refactor, all of these `On*` member functions are basically the same pattern, we can probably just pass in the method and switch on how to parse it and use the same member function for them all (or most of them).  Not as part of this PR, but just sharing.

File: components/brave_wallet/browser/eth_tx_controller.cc [hunk 2]
```diff
@@ -153,7 +153,83 @@ void EthTxController::AddUnapprovedTransaction(
     return;
   }
 
-  EthTxStateManager::TxMeta meta(std::make_unique<EthTransaction>(*tx));
+  auto tx_ptr = std::make_unique<EthTransaction>(*tx);
+
+  // Use default gas price and limit for ETHSend.
+  if (tx_ptr->data().empty()) {  // ETHSend
+    if (!tx_ptr->gas_limit())
+      tx_ptr->set_gas_limit(21000);
```
Comment (bbondy) [inline components/brave_wallet/browser/eth_tx_controller.cc:161]: Pls add a constant for this value in `components/brave_wallet/browser/brave_wallet_constants.h`
`kDefaultSendEthGasLimit` or something like that

File: components/brave_wallet/browser/eth_tx_controller.cc [hunk 3]
```diff
@@ -153,7 +153,83 @@ void EthTxController::AddUnapprovedTransaction(
     return;
   }
 
-  EthTxStateManager::TxMeta meta(std::make_unique<EthTransaction>(*tx));
+  auto tx_ptr = std::make_unique<EthTransaction>(*tx);
+
+  // Use default gas price and limit for ETHSend.
+  if (tx_ptr->data().empty()) {  // ETHSend
```
Comment (bbondy) [inline components/brave_wallet/browser/eth_tx_controller.cc:159]: Might be nicer to call `GetTransactionInfoFromData` from `components/brave_wallet/browser/eth_data_parser.h` and check if the tx_type is ETHSend from this mojo type:
```
enum TransactionType {
  ETHSend,
  ERC20Transfer,
  ERC20Approve,
  Other
};
```
You could allow `GetTransactionInfoFromData` to take in nullptr for the parameter info when not needed and don't parse it in that case.

File: components/brave_wallet/browser/eth_tx_controller.cc [hunk 4]
```diff
@@ -153,7 +153,83 @@ void EthTxController::AddUnapprovedTransaction(
...
+    if (!tx_ptr->gas_limit())
+      tx_ptr->set_gas_limit(21000);
+    if (!tx_ptr->gas_price())
+      tx_ptr->set_gas_price(150 * 1e9);  // 150 Gwei
```
Comment (bbondy) [inline components/brave_wallet/browser/eth_tx_controller.cc:163]: does this value overflow? I don't think it'll be treated as uint256_t, maybe you need `150ULL`? Maybe a constant for this value too.

File: components/brave_wallet/browser/eth_tx_controller_unittest.cc [hunk 5]
```diff
@@ -5,13 +5,310 @@
...
+    keyring_controller_.reset(new KeyringController(&prefs_));
+
+    auto tx_state_manager = std::make_unique<EthTxStateManager>(
+        &prefs_, rpc_controller_->MakeRemote());
+    auto nonce_tracker = std::make_unique<EthNonceTracker>(
+        tx_state_manager.get(), rpc_controller_.get());
+    auto pending_tx_tracker = std::make_unique<EthPendingTxTracker>(
+        tx_state_manager.get(), rpc_controller_.get(), nonce_tracker.get());
+
+    eth_tx_controller_.reset(new EthTxController(
```
Comment (bbondy) [inline components/brave_wallet/browser/eth_tx_controller_unittest.cc:106]: You can use factories to get all of these and no need to create the intermediate things like `nonce_tracker` manually in the test.  See `browser/brave_wallet/brave_wallet_provider_impl_unittest.cc` for an example.

File: components/brave_wallet/browser/eth_tx_controller_unittest.cc [hunk 6]
```diff
@@ -5,13 +5,310 @@
...
+      base::BindOnce(&AddUnapprovedTransactionSuccessCallback, &callback_called,
+                     &tx_meta_id));
+
+  base::RunLoop().RunUntilIdle();
+  EXPECT_TRUE(callback_called);
+  auto tx_meta = eth_tx_controller_->GetTxForTesting(tx_meta_id);
+  EXPECT_TRUE(tx_meta);
+
+  // Default value will be used.
+  EXPECT_EQ(tx_meta->tx->gas_price(), 150u * 1e9);
```
Comment (bbondy) [inline components/brave_wallet/browser/eth_tx_controller_unittest.cc:307]: maybe ull here and below?

**Review by bbondy — CHANGES_REQUESTED** [review]

File: components/brave_wallet/browser/eth_tx_controller_unittest.cc (hunk 6 above)
Comment (yrliou) [reply components/brave_wallet/browser/eth_tx_controller_unittest.cc]: Use the defined constant now in https://github.com/brave/brave-core/pull/10131/commits/642734ae3d2a7c4e088c8274d06d34db7bbd2986.

File: components/brave_wallet/browser/eth_tx_controller.cc (hunk 4 above)
Comment (yrliou) [reply components/brave_wallet/browser/eth_tx_controller.cc]: Fixed in https://github.com/brave/brave-core/pull/10131/commits/642734ae3d2a7c4e088c8274d06d34db7bbd2986

File: components/brave_wallet/browser/eth_tx_controller.cc (hunk 3 above)
Comment (yrliou) [reply components/brave_wallet/browser/eth_tx_controller.cc]: Fixed in https://github.com/brave/brave-core/pull/10131/commits/642734ae3d2a7c4e088c8274d06d34db7bbd2986

File: components/brave_wallet/browser/eth_tx_controller.cc (hunk 2 above)
Comment (yrliou) [reply components/brave_wallet/browser/eth_tx_controller.cc]: Fixed in https://github.com/brave/brave-core/pull/10131/commits/642734ae3d2a7c4e088c8274d06d34db7bbd2986

File: components/brave_wallet/browser/eth_tx_controller_unittest.cc (hunk 5 above)
Comment (yrliou) [reply components/brave_wallet/browser/eth_tx_controller_unittest.cc:107]: It seems I would need to create my own rpc_controller and hence all these classes to use the desired url_loader_factory to intercept calls?

File: components/brave_wallet/browser/eth_json_rpc_controller.cc (hunk 1 above)
Comment (yrliou) [reply components/brave_wallet/browser/eth_json_rpc_controller.cc:529]: Agreed, created https://github.com/brave/brave-browser/issues/18188

File: components/brave_wallet/browser/eth_tx_controller_unittest.cc (hunk 5 above)
Comment (bbondy) [reply components/brave_wallet/browser/eth_tx_controller_unittest.cc:107]: ok you can leave as is in that case

**Review by bbondy — APPROVED** [review]

**Pull request merged**